// meteor-publish-join/dist/helpers.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

exports.isObject = isObject;
exports.isString = isString;
exports.isNumber = isNumber;
exports.isFunction = isFunction;
exports.isUndefined = isUndefined;
exports.throwError = throwError;
exports.isBoolean = isBoolean;
function isObject(object) {
  return (typeof object === 'undefined' ? 'undefined' : _typeof(object)) === 'object' && !!object;
}

function isString(str) {
  return Object.prototype.toString.call(str) === '[object String]';
}

function isNumber(number) {
  return Object.prototype.toString.call(number) === '[object Number]';
}

function isFunction(func) {
  return Object.prototype.toString.call(func) === '[object Function]';
}

function isUndefined(value) {
  return value === undefined;
}

function throwError(mes) {
  throw new Error(mes);
}

function isBoolean(obj) {
  return obj === true || obj === false || toString.call(obj) === '[object Boolean]';
}
